### PR TITLE
updated unittests for metadatatools

### PR DESCRIPTION
**Pull request by delarocamcas**
files changed:
test_metadata_tools.py: updated unittests for read and write methods , added helper function and method to verify tags on EXIF_constants

### CODE REVIEW

File: metadata_tools.py
Comment (delarocamcas) [inline metadata_tools.py:1]: base timeout is no longer compatible with python 3.10

File: tests/__init__.py
Comment (delarocamcas) [inline tests/__init__.py:1]: added init for pytest

File: tests/test_metadata_tools.py [hunk 1]
```diff
@@ -2,26 +2,76 @@
...
 
     def test_write_exif_tags(self):
-        pass
+        """tests exif attach function"""
+        exif_dict = {'EXIF:LensMake': 'Samsung', 'IPTC:City': 'San Boston', 'EXIF:ApertureValue': '1.5'}
+        self.md.write_exif_tags(exif_dict=exif_dict)
+        exif_return = self.md.read_exif_tags()
+        self.assertEqual("Samsung", exif_return['EXIF:LensMake'])
+        self.assertEqual('San Boston', exif_return['IPTC:City'])
+        self.assertEqual('1.5', exif_return['EXIF:ApertureValue'])
```
Comment (delarocamcas) [inline tests/test_metadata_tools.py:61]: tests whether all exif tags in EXIF constants base class are acceptable to exiftool.

File: base_constants.py [hunk 2]
```diff
@@ -8,13 +8,23 @@ def iterate_constants(cls):
             if not key.startswith("__") and not callable(value):
                 yield key, value
```
Comment (delarocamcas) [inline base_constants.py:10]: adds function to base that allows is_valid_constant to check entire dictionary.

File: metadata_tools.py [hunk 3]
```diff
@@ -41,10 +41,15 @@ def read_exif_tags(self):
     def write_exif_tags(self, exif_dict):
         """Writes all exif tags to an image with a single call to ExifTool"""
         self.logger.info(f"Processing EXIF data for: {self.path}")
+
```
Comment (delarocamcas) [inline metadata_tools.py:44]: if any value return by dictionary is not valid , will raise error and return booleans of which terms are not valid.

File: tests/test_metadata_tools.py [hunk 4]
```diff
@@ -2,26 +2,76 @@
...
+
+            # Check for IFD0 group
+            if table.get('name') == 'IFD0':
+                for tag in table.findall('tag'):
+                    tag_name = f"IFD0:{tag.get('name')}"
+                    tags.add(tag_name)
+
+        return tags
+
+
```
[the remark this answers is not in the document]
Comment (delarocamcas) [reply tests/test_metadata_tools.py:39]: Let me know on this latest round of review if the helper function and unittest of checking the EXIF_constants.py terms if they are valid against exiftools is redundant or will rarely if ever be useful, if so I can delete and rebase.

File: tests/test_metadata_tools.py [hunk 5]
```diff
@@ -1,27 +1,92 @@
...
+        exif_dict = {'EXIF:TEST1': 'Samsung', 'IPTC:CopyrightNotice': 'CaliforniaAcademy', 'EXIF:ApertureValid': '1.5'}
+        with self.assertRaises(ValueError) as context:
+            self.md.write_exif_tags(exif_dict=exif_dict)
+
+        self.assertEqual(str(context.exception),
+                         "Invalid keys in exif_dict, check exif "
+                         "constants:{'EXIF:TEST1': False, 'IPTC:CopyrightNotice': True, 'EXIF:ApertureValid': False}")
+
 
     def test_accepted_exif_tags(self):
```
Comment (delarocamcas) [inline tests/test_metadata_tools.py:77]: tests whether write exif will reject inccorrect exif tags and return correct error.

File: tests/test_metadata_tools.py [hunk 6]
```diff
@@ -1,27 +1,92 @@
...
+        datetime_test = datetime_test.strftime('%Y:%m:%d %H:%M:%S.%f')
+        exif_dict = {"XMP:CreatorCity": 'San Francisco', 'IPTC:CopyrightNotice': 'CaliforniaAcademy',
+                     'XMP:CreateDate': f'{datetime_test}'}
+        self.md.write_exif_tags(exif_dict=exif_dict)
+        exif_return = self.md.read_exif_tags()
+        print(exif_return)
+        self.assertEqual("San Francisco", exif_return['XMP:CreatorCity'])
+        self.assertEqual('CaliforniaAcademy', exif_return['IPTC:CopyrightNotice'])
+        self.assertEqual(str(f'{datetime_test}'), str(exif_return['XMP:CreateDate']))
+
```
Comment (delarocamcas) [inline tests/test_metadata_tools.py:66]: tests whether write exif will reject inccorrect exif tags and return correct error.

File: tests/test_metadata_tools.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (foozleface) [reply tests/test_metadata_tools.py:39]: probably rarely useful. But if the code works and is tested, ok to leave.

**Review by foozleface — APPROVED** [review]

**Comment by foozleface** [comment]
not landing untill you decide whether you want to retain the tag-export routine commented on above.

**Pull request merged**